serializers/avro/src/main/java/io/pravega/schemaregistry/serializer/avro/impl/AvroSerializerFactory.java

```diff
@@ -0,0 +1,188 @@
+                .entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey,
+                        x -> new AvroSerializer<>(groupId, schemaRegistryClient, x.getValue(), config.getEncoder(),
+                                config.isRegisterSchema())));
+        return new MultiplexedSerializer<>(serializerMap);
+    }
+
+    /**
+     * A multiplexed Avro Deserializer that takes a map of schemas and deserializes events into those events depending
+     * on the object type information in {@link EncodingInfo}.
+     *
+     * @param config  Serializer config.
+     * @param schemas map of avro schemas.
+     * @param <T>     Base type of schemas.
+     * @return a Deserializer which can deserialize events of different types in the stream into typed objects.
+     */
+    public static <T> Serializer<T> multiTypeDeserializer(
+            SerializerConfig config, Map<Class<? extends T>, AvroSchema<T>> schemas) {
+        Preconditions.checkNotNull(config);
+        Preconditions.checkNotNull(schemas);
+        Preconditions.checkArgument(config.isWriteEncodingHeader(), "Events should be tagged with encoding ids.");
+
+        String groupId = config.getGroupId();
+        SchemaRegistryClient schemaRegistryClient = initForDeserializer(config);
+
+        EncodingCache encodingCache = new EncodingCache(groupId, schemaRegistryClient);
+
+        Map<String, AbstractDeserializer<T>> deserializerMap = schemas
+                .values().stream().collect(Collectors.toMap(x -> x.getSchemaInfo().getType(),
+                        x -> new AvroDeserializer<>(groupId, schemaRegistryClient, x, config.getDecoders(), encodingCache)));
+        return new MultiplexedDeserializer<>(groupId, schemaRegistryClient, deserializerMap, config.getDecoders(),
+                encodingCache);
+    }
+
+    /**
+     * A multiplexed Avro Deserializer that takes a map of schemas and deserializes events into those events depending
+     * on the object type information in {@link EncodingInfo}.
+     *
+     * @param config  Serializer config.
+     * @param schemas map of avro schemas.
+     * @param <T>     Base type of schemas.
+     * @return a Deserializer which can deserialize events of different types in the stream into typed objects or a generic
+     * object
+     */
+    public static <T> Serializer<Either<T, Object>> typedOrGenericDeserializer(
+            SerializerConfig config, Map<Class<? extends T>, AvroSchema<T>> schemas) {
+        Preconditions.checkNotNull(config);
+        Preconditions.checkNotNull(schemas);
+        Preconditions.checkArgument(config.isWriteEncodingHeader(), "Events should be tagged with encoding ids.");
+
+        String groupId = config.getGroupId();
+        SchemaRegistryClient schemaRegistryClient = initForDeserializer(config);
+
+        EncodingCache encodingCache = new EncodingCache(groupId, schemaRegistryClient);
+
+        Map<String, AbstractDeserializer<T>> deserializerMap = schemas
```

Comment:
Much of the code in this method looks the same as that in `multiTypeDeserializer(..)`. Could we extract a method and reuse it?

Reply:
i can extract the deserializer map generation into a separate method. 
the lines above it are common inits that everyone does -- initDeserializer and create encoding cache. I could have moved creation of encoding cache inside the initDeserializer too but that would have meant returning a complex object that encapsulated the tuple - schemaregisrtyclient and encodingcache. 
so kept is simpler instead.